### DISCUSSION TITLE
Add binding for SOAP 1.1 and SOAP 1.2 for same endpoint address

### DISCUSSION THREAD

**Discussion by danreg** [CoreWCF/CoreWCF#1375] [Q&A]
I need to create a SOAP Interface which offers methods for SOAP 1.1 and SOAP 1.2 at the same endpoint address.

When I import the old net framework based WSDL in SOAPUI I'm getting this 
![image](https://github.com/CoreWCF/CoreWCF/assets/16964633/e3e6be8f-2ced-40ad-b7ff-106af213f097)

So to get this working with CoreWCF I tried to generate a custom binding, but I think there is some mistake.

`serviceBuilder.AddServiceEndpoint<InterfaceService, IInterfaceService>(GetBindingForEndpoint(), "/int/interfaceservice.asmx");`

` private static Binding GetBindingForEndpoint()
        {
            CustomBinding result = new CustomBinding();

            TextMessageEncodingBindingElement textBindingElement1 = new TextMessageEncodingBindingElement();
            textBindingElement1.MessageVersion = MessageVersion.CreateVersion(EnvelopeVersion.Soap11, AddressingVersion.None);
            result.Elements.Add(textBindingElement1);

            TextMessageEncodingBindingElement textBindingElement = new TextMessageEncodingBindingElement();
            textBindingElement.MessageVersion = MessageVersion.CreateVersion(EnvelopeVersion.Soap12, AddressingVersion.None);
            result.Elements.Add(textBindingElement);

            HttpTransportBindingElement httpBindingElement = new HttpTransportBindingElement();
            httpBindingElement.MaxBufferSize = int.MaxValue;
            httpBindingElement.MaxReceivedMessageSize = int.MaxValue;
            result.Elements.Add(httpBindingElement);
            return result;

        }`

@mconnew

**Comment by mconnew** [comment]
By default WCF/CoreWCF doesn't support what you are trying to do out of the box as the binding is tied to the incoming request path. It looks at the HTTP request path and sends it to a component with a configured message encoder. The only ability to switch comes later in the pipeline after the message had been decoded. This means the only way to have multiple endpoints configured at the same uri is if they all use the same binding. There are 2 potential ways to achieve what you are doing tho.  

1. Fake a different path for each endpoint
CoreWCF is just middleware built on top of asp.net core. You can place some inline middleware before the call to `app.UseServiceModel` which looks at the content-type of the incoming request (SOAP1.1 and SOAP1.2 use different content-type values). Based on the content type, modify the request path to something different, eg append /soap11 or /soap12. Configure CoreWCF with 2 different service endpoints at different listen paths. CoreWCF will see things as if clients are using 2 different paths, but the clients will be using a single path. This will make a WSDL have the wrong paths though.  

2. Create a custom implementation of MessageEncodingBindingElement
Create an implementation which wraps two different implementations of TextMessageEncodingBindingElement. You will likely need to implement a wrapping MessageEncoderFactory and MessageEncoder as MessageEncoder is where you would choose between the two implementations. There's a few nuances you'll need to worry about such as having IsContentTypeSupported try the content type with potentially both inner encoders. In each of the ReadMessage method variants you'll need to fetch the content type from the HttpRequestMessageProperty found in the Message.Properties collection. You'll need some way to tell the encoder on the Write path which of the two to use. You can do this by adding a custom message property stating which encoder you used (could just be a string with either SOAP11 or SOAP12). Then you'll need a custom IDispatchMessageInspector which returns this encoder value from AfterReceiveRequest, then in BeforeSendReply, adds it to the outgoing Message properties so that in WriteMessage you can retrieve the message property to know which encoder to use for the reply.

Option 2 is the "cleanest" solution as it's less of a hack, although again your WSDL will be weird as it won't have the binding info available. You could implement IWsdlExportExtension, but I honestly have no idea how your WSDL would look with one endpoint but two different soap versions. If it's specified as two separate endpoints which happen to have the same endpoint address, then there's a lot of work you would have to do. One option you could do is if your contract is never going to change, take your existing wsdl and host it in ASP.NET Core as static files and skip having CoreWCF generate it.  

Option 1 is the easiest solution to implement. Again depending how the wsdl expresses two different bindings on the same url, there's a good chance it's the easiest to fix up the wsdl if you aren't going to statically host it as files. You might be able to provide a custom IServiceBehavior (which changes nothing about the service) which also implements IWsdlExportExtension which modifies the service endpoint url's to be the same.

**Comment by danreg** [comment]
@mconnew thx for your detailed answer. Unfortunetaly no easy solution.
Just as a note, the current WSDL has two bindings which look like this

![image](https://github.com/CoreWCF/CoreWCF/assets/16964633/64d07774-e638-4c7a-a460-776b5eb9b8ac)



I will think about a solution :)

I tried to add 
```
serviceBuilder.AddServiceEndpoint<InterfaceService, IInterfaceService>(new BasicHttpBinding(), "/int/interfaceservice.asmx")
serviceBuilder.AddServiceEndpoint<InterfaceService, IInterfaceService>(new BasicHttpBinding(), "/int/interfaceservice12.asmx"
```

for a basic test, but that does not work, I got http 400 for the second endpoint

**Comment by john-surcombe** [comment]
Option 1 hack in my case looked like this:

```
app.Use(async (context, next) =>
{
    if (context.Request.Path.StartsWithSegments("/API/Service.asmx"))
    {
        if (context.Request.ContentType?.StartsWith("application/soap+xml", StringComparison.OrdinalIgnoreCase) ?? false)
            context.Request.Path = context.Request.Path.Value!.Replace("/Service.asmx", "/Service12.asmx");
        else
            context.Request.Path = context.Request.Path.Value!.Replace("/Service.asmx", "/Service11.asmx");
    }

    await next();
});
```

This leaves you with two new canonical endpoints (Service11 and Service12), which are the ones in the WSDL, but just adds some backward compatibility for clients relying on the single legacy WCF endpoint.
